proposals/NNNN-importing-forward-declared-objc-interfaces-and-protocols.md

```diff
@@ -0,0 +1,177 @@
+@end
+
+ForwardDeclaredInterface *CFunctionReturningAForwardDeclaredInterface1();
+void CFunctionTakingAForwardDeclaredInterface1(
+    ForwardDeclaredInterface *param);
+
+NSObject<ForwardDeclaredProtocol> *CFunctionReturningAForwardDeclaredProtocol1();
+void CFunctionTakingAForwardDeclaredProtocol1(
+    id<ForwardDeclaredProtocol> param);
+```
+
+_Swift_
+```swift
+class IncompleteTypeConsumer1 : NSObject {
+  init!()
+}
+```
+We wish to make the experience of consuming an Objective-C API with forward declarations from Swift more consistent with the experience of consuming the API from Objective-C.
+
+## Proposed solution
+
+We propose the following representation for forward declared Objective-C interfaces and protocols in Swift:
+
+```swift
+// @class Foo turns into
+@available(*, unavailable, message: “This Objective-C class has only been forward declared; import its owning module to use it”)
+class Foo : NSObject {}
+
+// @protocol Bar turns into
+@available(*, unavailable, message: “This Objective-C protocol has only been forward declared; import its owning module to use it”)
+protocol Bar : NSObjectProtocol {}
+```
+
+The idea is to introduce the minimal change that will make Objective-C APIs usable in a predictable safe manner.
+
+The aforementioned Objective-C API with this change looks like this from Swift:
+
+```swift
+@available(*, unavailable, message: "This Objective-C class has only been forward-declared; import its owning module to use it")
+class ForwardDeclaredInterface {
+}
+@available(*, unavailable, message: "This Objective-C protocol has only been forward-declared; import its owning module to use it")
+protocol ForwardDeclaredProtocol : NSObjectProtocol {
+}
+class IncompleteTypeConsumer1 : NSObject {
+  var propertyUsingAForwardDeclaredProtocol1: ForwardDeclaredProtocol!
+  var propertyUsingAForwardDeclaredInterface1: ForwardDeclaredInterface!
+  init!()
+  func methodReturningForwardDeclaredProtocol1() -> ForwardDeclaredProtocol!
+  func methodReturningForwardDeclaredInterface1() -> ForwardDeclaredInterface!
+  func methodTakingAForwardDeclaredProtocol1(_ param: ForwardDeclaredProtocol!)
+  func methodTakingAForwardDeclaredInterface1(_ param: ForwardDeclaredInterface!)
+}
+func CFunctionReturningAForwardDeclaredInterface1() -> ForwardDeclaredInterface!
+func CFunctionTakingAForwardDeclaredInterface1(_ param: ForwardDeclaredInterface!)
+func CFunctionReturningAForwardDeclaredProtocol1() -> ForwardDeclaredProtocol!
+func CFunctionTakingAForwardDeclaredProtocol1(_ param: ForwardDeclaredProtocol!)
+```
+
+More usage examples can be found in these tests introduced here: https://github.com/apple/swift/pull/61606
+
+## Detailed design
+
+Modifications lie almost exclusively in the ClangImporter -- specifically in the SwiftDeclConverter. If asked to convert a
+`clang::ObjCInterfaceDecl` or `clang::ObjCProtocolDecl` with no definition the SwiftDeclConverter will now return a placeholder type instead of
```

Comment:
```suggestion
`clang::ObjCInterfaceDecl` or `clang::ObjCProtocolDecl` with no definition, `SwiftDeclConverter` will now return a placeholder type instead of
```